File: js/src/iterator.ts
/* Iterator — a cursor over container children (flat each / recursive tree). */

import * as D from "./decode.ts";
import { EachRow, eachIter, treeWalk } from "./iterate.ts";
import { Blob } from "./blob.ts";

export { EachRow };

export class Iterator {
  _blob: Blob;
  _base: string;
  _recursive: boolean;
  _rows: EachRow[];
  _cursor: number;
  _populated: boolean;

  constructor(blob: Blob, path: string = "$", recursive: boolean = false) {
    this._blob = blob;
    this._base = path ? path : "$";
    this._recursive = recursive;
    this._rows = [];
    this._cursor = -1;
    this._populated = false;
  }

  _populate(): void {
    if (this._populated) return;
    this._populated = true;
    this._rows = [];
    const a = this._blob.data();
    const n = a.length;
    if (n === 0) return;

    let iroot = 0;
    if (this._base !== "$") {
      const r = D.lookup(a, n, 0, this._base);
      if (r.rc !== D.RC_OK) return;
      iroot = r.iStart;
    }

    if (this._recursive) treeWalk(a, n, iroot, this._base, this._base, 0, this._rows);
    else this._rows = eachIter(a, n, iroot, this._base);
  }

  // ── C++-style cursor protocol ─────────────────────────────────────
  next(): boolean {
    this._populate();
    this._cursor += 1;
    return this._cursor < this._rows.length;
  }
  current(): EachRow {
    return this._rows[this._cursor];
  }
  reset(): void {
    this._cursor = -1;
  }

  // ── iterable protocol ─────────────────────────────────────────────
  rows(): EachRow[] {
    this._populate();
    return this._rows.slice();
  }
  [Symbol.iterator](): IterableIterator<EachRow> {
    this._populate();
    return this._rows.slice()[Symbol.iterator]();
  }
}

File: rust/src/iterator.rs
//! `Iterator` — a cursor over container children (flat `each` / recursive `tree`).

use crate::blob::Blob;
use crate::decode as d;
use crate::iterate::{each_iter, tree_walk, EachRow};

/// Iterate over a container's children.
///
/// Supports flat (`each`) and recursive (`tree`) modes, mirroring the SQLite
/// extension's `msgpack_each` / `msgpack_tree` table-valued functions. Use the
/// C++-style [`Iterator::next`] / [`Iterator::current`] cursor, collect
/// [`Iterator::rows`], or drive it as a standard Rust iterator.
pub struct Iterator<'a> {
    blob: &'a Blob,
    base: String,
    recursive: bool,
    rows: Vec<EachRow>,
    cursor: i64,
    populated: bool,
}

impl<'a> Iterator<'a> {
    pub fn new(blob: &'a Blob, path: &str, recursive: bool) -> Iterator<'a> {
        Iterator {
            blob,
            base: if path.is_empty() {
                "$".to_string()
            } else {
                path.to_string()
            },
            recursive,
            rows: Vec::new(),
            cursor: -1,
            populated: false,
        }
    }

    fn populate(&mut self) {
        if self.populated {
            return;
        }
        self.populated = true;
        self.rows.clear();
        let a = self.blob.data();
        let n = a.len();
        if n == 0 {
            return;
        }

        let mut iroot = 0;
        if self.base != "$" {
            let (rc, istart, _) = d::lookup(a, n, 0, &self.base);
            if rc != d::RC_OK {
                return;
            }
            iroot = istart;
        }

        if self.recursive {
            let base = self.base.clone();
            tree_walk(a, n, iroot, &base, &base, 0, &mut self.rows);
        } else {
            self.rows = each_iter(a, n, iroot, &self.base);
        }
    }

    // ── C++-style cursor protocol ─────────────────────────────────────
    #[allow(clippy::should_implement_trait)]
    pub fn next(&mut self) -> bool {
        self.populate();
        self.cursor += 1;
        (self.cursor as usize) < self.rows.len()
    }
    pub fn current(&self) -> &EachRow {
        &self.rows[self.cursor as usize]
    }
    pub fn reset(&mut self) {
        self.cursor = -1;
    }

    /// Collect all rows.
    pub fn rows(mut self) -> Vec<EachRow> {
        self.populate();
        self.rows
    }
}

impl<'a> IntoIterator for Iterator<'a> {
    type Item = EachRow;
    type IntoIter = std::vec::IntoIter<EachRow>;
    fn into_iter(mut self) -> Self::IntoIter {
        self.populate();
        self.rows.into_iter()
    }
}
